### PR TITLE
OrtAuthenticator : in .netrc, raise 'machine' precedence

### PR DESCRIPTION
**Pull request by nnobelis**
When parsing the .netrc, the entry with the 'machine'
name has higher precedence than the 'default' one.
Current implementation always wrongly returns the first entry.

Also : the regex has been moved to a static field because compilation is expensive.
Maybe it would be worthwhile to make getNetrcAuthentication() an instance function
of OrtAuthenticator ? Thus we could make the regex private.

Signed-off-by: Nicolas Nobelis <nicolas.nobelis@bosch.io>

### CODE REVIEW

**Comment by sschuberth** [comment]
> When parsing the .netrc, the entry with the 'machine'
> name has higher precedence than the 'default' one.

From reading this, it's unclear to me whether you describe the behavior before your change or after your change. Could you please reword this to be ore clear?

> Current implementation always wrongly returns the first entry.

What is "first entry" here? The first entry in the file? The first matching entry?

**Review by sschuberth — CHANGES_REQUESTED** [review]
Please add a test case to `OrtAuthenticatorTest` to demonstrate the issue.

File: utils/src/test/kotlin/OrtAuthenticatorTest.kt [hunk 1]
```diff
@@ -61,6 +61,19 @@ class OrtAuthenticatorTest : WordSpec({
             authentication.password shouldBe "bar".toCharArray()
         }
 
+        "machine statement precedence" {
```
Comment (sschuberth) [inline utils/src/test/kotlin/OrtAuthenticatorTest.kt:64]: Please name test cases to that they form a sentence with the outer text. Here the outer text is "getNetrcAuthentication() should ..." but "machine statement precedence" does not complete that sentence in a meaningful / grammatically correct way.

Comment (sschuberth) [reply utils/src/test/kotlin/OrtAuthenticatorTest.kt:64]: Proposal: "prefer machine-specific entries over the default"

File: utils/src/main/kotlin/OrtAuthenticator.kt [hunk 2]
```diff
@@ -127,20 +127,29 @@ fun getNetrcAuthentication(contents: String, machine: String): PasswordAuthentic
 
     val iterator = lines.joinToString(" ").split(Regex("\\s+")).iterator()
 
-    var machineFound = false
     var login: String? = null
     var password: String? = null
+    var lastEntry: String? = null
```
Comment (sschuberth) [inline utils/src/main/kotlin/OrtAuthenticator.kt:132]: Maybe a better name: `currentMachine`

File: utils/src/main/kotlin/OrtAuthenticator.kt [hunk 3]
```diff
@@ -127,20 +127,29 @@ fun getNetrcAuthentication(contents: String, machine: String): PasswordAuthentic
...
-            "password" -> password = if (machineFound && iterator.hasNext()) iterator.next() else null
-            "default" -> machineFound = true
+            "machine" -> lastEntry = iterator.next()
+            "login" -> login = iterator.next()
+            "password" -> password = iterator.next()
+            "default" -> lastEntry = "default"
         }
 
-        if (login != null && password != null) return PasswordAuthentication(login, password.toCharArray())
+        if (login != null && password != null && lastEntry != null) {
```
Comment (sschuberth) [inline utils/src/main/kotlin/OrtAuthenticator.kt:143]: Nit: Maybe check `lastEntry != null` first to make the order nicely match the order in the line below for readability.

File: utils/src/main/kotlin/OrtAuthenticator.kt [hunk 4]
```diff
@@ -127,20 +127,29 @@ fun getNetrcAuthentication(contents: String, machine: String): PasswordAuthentic
...
+            credentialsPerMachine[lastEntry] = login to password
+        }
     }
 
+    credentialsPerMachine[machine]?.let {
+        return PasswordAuthentication(it.first, it.second.toCharArray())
+    }
+    credentialsPerMachine["default"]?.let {
+        return PasswordAuthentication(it.first, it.second.toCharArray())
+    }
```
Comment (sschuberth) [inline utils/src/main/kotlin/OrtAuthenticator.kt:153]: How about this instead:

    return listOfNotNull(credentialsPerMachine[machine], credentialsPerMachine["default"]).firstOrNull()
        ?.let { PasswordAuthentication(it.first, it.second.toCharArray()) }

**Review by sschuberth — CHANGES_REQUESTED** [review]
BTW, thanks for this fix, good finding in general!

File: utils/src/main/kotlin/OrtAuthenticator.kt (hunk 4 above)
Comment (nnobelis) [reply utils/src/main/kotlin/OrtAuthenticator.kt:153]: The solution before was doing only one map lookup in the best of case and yours is always doing two.
I know we are talking about micro optimizations here but nevertheless, we should not do useless things.

**Comment by Martin-Idel-SI** [comment]
@nnobelis Could we maybe add some logging for finding passwords (maybe info-level)? Obviously we don't log the password, but we could log if an entry is found and used in the .netrc file and equivalently if a password is found elsewhere.

This would make debugging permission issues somewhat easier since credentials are taken from several sources already.

File: utils/src/main/kotlin/OrtAuthenticator.kt (hunk 4 above)
Comment (nnobelis) [reply utils/src/main/kotlin/OrtAuthenticator.kt:153]: how about
```
credentialsPerMachine[machine]?.let {
        return PasswordAuthentication(it.first, it.second.toCharArray())
}
return credentialsPerMachine["default"]?.let {
        PasswordAuthentication(it.first, it.second.toCharArray())
}
```

Comment (sschuberth) [reply utils/src/main/kotlin/OrtAuthenticator.kt:153]: Well, my main aim was to reduce the total amount of code, and in particular to not have the duplication to construct `PasswordAuthentication`.

> The solution before was doing only one map lookup in the best of case and yours is always doing two.
I know we are talking about micro optimizations here but nevertheless, we should not do useless things.

I basically like that line of thinking 😀 But that could be easily fixed by

    return sequenceOf(credentialsPerMachine[machine], credentialsPerMachine["default"]).find { it != null }
        ?.let { PasswordAuthentication(it.first, it.second.toCharArray()) }

Comment (nnobelis) [reply utils/src/main/kotlin/OrtAuthenticator.kt]: But is it not the same ?
When writing `sequenceOf(credentialsPerMachine[machine], credentialsPerMachine["default"])` he always do two lookups first even if later, only one is consumed by `find`.

Comment (nnobelis) [reply utils/src/main/kotlin/OrtAuthenticator.kt]: Ok I found the solution
```
return (credentialsPerMachine[machine] ?: credentialsPerMachine["default"])?.let {
        PasswordAuthentication(it.first, it.second.toCharArray())
    }
```
but then you have no specific logging. Please advise.

Comment (sschuberth) [reply utils/src/main/kotlin/OrtAuthenticator.kt]: No, that's exactly the difference between a sequence and a list: Sequence members are evaluated lazily. So if the first entry already is non-null, the second entry will not even be looked at. Also see e.g. https://www.baeldung.com/kotlin/sequences.

File: utils/src/main/kotlin/OrtAuthenticator.kt [hunk 5]
```diff
@@ -146,9 +147,11 @@ fun getNetrcAuthentication(contents: String, machine: String): PasswordAuthentic
     }
 
     credentialsPerMachine[machine]?.let {
+        OrtAuthenticator.logger().debug() { "Found .netrc entry for $machine" }
```
Comment (sschuberth) [inline utils/src/main/kotlin/OrtAuthenticator.kt:150]: Please use `OrtAuthenticator.log.debug { ... }` and a dot at the end of log message sentences.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: utils/src/main/kotlin/OrtAuthenticator.kt (hunk 4 above)
Comment (sschuberth) [reply utils/src/main/kotlin/OrtAuthenticator.kt]: > but then you have no specific logging. Please advise.

I made up my mind. let's not over-complicate things by using sequences for only two elements here, but basically use your code as-is. I just have one more formatting nit (in a separate comment).

File: utils/src/main/kotlin/OrtAuthenticator.kt [hunk 6]
```diff
@@ -127,20 +127,28 @@ fun getNetrcAuthentication(contents: String, machine: String): PasswordAuthentic
...
+        if (currentMachine != null && login != null && password != null) {
+            credentialsPerMachine[currentMachine] = login to password
+        }
     }
 
-    return null
+    credentialsPerMachine[machine]?.let {
+        return PasswordAuthentication(it.first, it.second.toCharArray())
+    }
+    return credentialsPerMachine["default"]?.let {
```
Comment (sschuberth) [inline utils/src/main/kotlin/OrtAuthenticator.kt:153]: Nit: Blank line before this one.

File: utils/src/main/kotlin/OrtAuthenticator.kt (hunk 4 above)
Comment (nnobelis) [reply utils/src/main/kotlin/OrtAuthenticator.kt]: I think there is a misunderstanding here : yes, because this is a sequence, the .find is going to be called only once.
However, because of Kotlin eager evaluation, when calling sequenceOf, it is going to evaluate both two parameters and therefore going to do two lookups.
https://ahsensaeed.com/kotlin-lazy-evaluation-haskell-laziness/

I don't see any lazy in the parameters of `sequenceOf`so I assume all the parameters are evaluated when `sequenceof` is called.

File: utils/src/main/kotlin/OrtAuthenticator.kt [hunk 7]
```diff
@@ -26,6 +26,7 @@ import java.net.Proxy
 import java.net.ProxySelector
 
 import org.apache.logging.log4j.Level
+import org.apache.logging.log4j.kotlin.logger
```
Comment (sschuberth) [inline utils/src/main/kotlin/OrtAuthenticator.kt:29]: Commit message: s/add/Add/, no dot at the end of a title

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: utils/src/main/kotlin/OrtAuthenticator.kt (hunk 4 above)
Comment (sschuberth) [reply utils/src/main/kotlin/OrtAuthenticator.kt]: You're right in that my usage of `sequenceOf()` was wrong, I should have used `sequence {}` as in [this example](https://pl.kotl.in/dPdSKTM1t).

File: utils/src/main/kotlin/OrtAuthenticator.kt [hunk 8]
```diff
@@ -127,20 +127,29 @@ fun getNetrcAuthentication(contents: String, machine: String): PasswordAuthentic
 
     val iterator = lines.joinToString(" ").split(Regex("\\s+")).iterator()
 
-    var machineFound = false
     var login: String? = null
     var password: String? = null
+    var currentMachine: String? = null
```
Comment (sschuberth) [inline utils/src/main/kotlin/OrtAuthenticator.kt:132]: Sorry, the commit message nit to change "in" to In" after ":" is still valid.

**Review by sschuberth — CHANGES_REQUESTED** [review]

**Review by sschuberth — APPROVED** [review]
